authorization/src/main/kotlin/org/modelix/authorization/permissions/SchemaInstance.kt

```diff
@@ -0,0 +1,180 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.authorization.permissions
+
+/**
+ * Instantiates the abstract schema with data from an actual system.
+ */
+class SchemaInstance(val schema: Schema) {
+    /**
+     * Also contains child instances
+     */
+    val resources: MutableMap<ResourceInstanceReference, ResourceInstance> = LinkedHashMap()
+
+    init {
+        // resources without parameters can be instantiated directly without waiting for additional information
+        schema.resources.filter { it.value.parameters.isEmpty() }.forEach {
+            instantiateResource(ResourceInstanceReference(it.key, emptyList(), null))
+        }
+    }
+
+    fun getAllPermissions() = resources.values.flatMap { it.permissions.values }
+
+    fun instantiateResource(ref: ResourceInstanceReference): ResourceInstance {
+        resources[ref]?.let { return it }
+
+        val parentInstance = ref.parent?.let { instantiateResource(it) }
+        val resourceSchema = requireNotNull((parentInstance?.resourceSchema?.resources ?: schema.resources)[ref.name]) {
+            "Resource not found: ${ref.name}"
+        }
+        val newInstance = ResourceInstance(
+            parentInstance,
+            resourceSchema,
+            ref,
+        )
+        resources[ref] = newInstance
+        parentInstance?.let { it.childResources[ref] = newInstance }
+        resourceSchema.permissions.forEach { newInstance.getOrCreatePermissionInstance(it.key) }
+
+        // resources without parameters can be instantiated directly without waiting for additional information
+        resourceSchema.resources.filter { it.value.parameters.isEmpty() }.forEach {
+            instantiateResource(ResourceInstanceReference(it.key, emptyList(), ref))
+        }
+
+        updateIncludes()
+        return newInstance
+    }
+
+    private fun evaluateExpression(expr: IExpression, source: ResourceInstance): String {
+        return when (expr) {
+            is AddPrefix -> expr.prefix + evaluateExpression(expr.expr, source)
+            is SourceParameterValue -> source.reference.parameterValues[source.resourceSchema.parameters.indexOf(expr.name)]
+        }
+    }
+
+    fun instantiatePermission(ref: PermissionInstanceReference): ResourceInstance.PermissionInstance {
+        return instantiateResource(ref.resource)
+            .getOrCreatePermissionInstance(ref.permissionName)
+    }
+
+    fun updateIncludes() {
+        resources.values.forEach { it.updateIncludes() }
+    }
+
+    inner class ResourceInstance(
+        val parent: ResourceInstance?,
+        val resourceSchema: Resource,
+        val reference: ResourceInstanceReference,
+    ) {
+        val childResources: MutableMap<ResourceInstanceReference, ResourceInstance> = LinkedHashMap()
+        val relationTargetInstances: MutableMap<String, ResourceInstance> = LinkedHashMap()
```

Comment:
`relationTargetInstances` is always empty

Reply:
Removed.